### PR TITLE
feat: Add ValidationPolicy & ValidationEvent

### PR DESCRIPTION
**Pull request by karlkfi**
- Add ValidationPolicy:
  ExitEarly (default) - exit before apply/prune if any objects are invalid
  SkipInvalid - apply/prune valid objects & skip invalid ones
- Add ValidationEvent to be sent once for every invalid object or set of
  objects, when the SkipInvalid policy is selected. For ExitEarly,
  return an error for reverse compatibility.
- Add validation.Collector to simplify aggregating validation errors
  from multiple sources and extracting invalid object IDs.
- Add invalid objects to the TestContext so they can be retained in
  the inventory (only if already present). This primarily applies to
  invalid annotations and dependencies. Objects without name or kind
  should never be added to the inventory.
- Update Solver to use validation.Collector and filter invalid objects.
- Add e2e test for invalid objects.
- Update Printers to handle ValidationEvent
- Add ExternalDependencyError & InvalidAnnotationError to make it easier
  to handle and introspect validation errors.

fixes: https://github.com/kubernetes-sigs/cli-utils/issues/484

### CODE REVIEW

File: pkg/object/validator/error.go
Comment (haiyanmeng) [inline pkg/object/validator/error.go:30]: `a resources` -> `a resource`

File: pkg/object/validator/validate.go
Comment (haiyanmeng) [inline pkg/object/validator/validate.go:99]: Should we validate [CyclicDependencyError](https://github.com/kubernetes-sigs/cli-utils/blob/45a0b5b7e6b292c0a59c899aa50a3662a0c1cf7d/pkg/object/graph/graph.go#L139) here too?

File: pkg/object/validator/validate.go
Comment (haiyanmeng) [inline pkg/object/validator/validate.go:112]: Why duplicates are not allowed here?

File: pkg/apply/applier.go
Comment (haiyanmeng) [inline pkg/apply/applier.go:202]: Does the order of these filters matter?

File: pkg/apply/applier.go
Comment (haiyanmeng) [inline pkg/apply/applier.go:192]: Does the order of these filters matter?

File: pkg/object/validator/policy.go
Comment (mortent) [inline pkg/object/validator/policy.go:7]: So it seems like `SkipInvalid` is the approach that best fits our overall approach to errors here. It essentially follows the "continue-on-error" direction, but also avoids applying resources that we know there are problems with. I think there will be some issues that will prevent us from continuing (name/namespace errors comes to mind. Also, circular dependencies isn't currently caught by the validation and it is not clear how we would skip resources in that situation).
Could we simplify by making this the behavior until we have evidence that something else is needed?


I'm not convinced there is much need for the `ApplyAll` policy here, as it seems to suggest we'll be applying resources that we know have problems and effectively ignoring any annotations that can't be parsed.

I think `ExitEarly` do have some good use-cases, but I'm uncertain if this is the right way to do it. For most of the validation, we want to let users shift it left and detect these issues in their CI/CD pipelines rather than when they are ready to apply. So it seems like we want this to be available outside of a pre-step to actually applying it. Also, we already provide a dry-run flag, which should detect these issues as well as many additional validation errors.

File: pkg/object/graph/depends.go
Comment (mortent) [inline pkg/object/graph/depends.go:100]: Could we handle the validation here where it is needed, rather than doing all validation early? I realize doing it all early is beneficial for detecting all errors early and then report them all in an early exit, but it seems like some validation errors will only be caught later. Circular dependencies would be one of them.

File: pkg/apply/applier.go
Comment (mortent) [inline pkg/apply/applier.go:149]: It seems like errors in name and/or namespace would be somewhat different than other validation errors. They will have to be detected before we update the inventory. Skipping those means we would make sure they are not added to the inventory, but also make sure we don't end up accidentally pruning any resources that have been skipped.

File: pkg/object/validator/validate.go
Comment (karlkfi) [reply pkg/object/validator/validate.go:99]: In order to find a cyclic error, we'd have to build the graph first. It seems pretty inefficient to do that twice. I'm not sure if we want to move graph building forward in front of validation... and building the graph might require valid inputs...

I guess we could make the graph SortObj emit events, but it doesn't do that now so it would require some significant changes...

File: pkg/object/validator/validate.go
Comment (karlkfi) [reply pkg/object/validator/validate.go:112]: If the user put two of the same dependencies in the same depends-on annotation, it's probably a mistake, right?

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go:202]: They get executed in this order. Filtering invalid objects first prevents the other filters from erroring due to an invalid object.

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go:192]: They get executed in this order. Filtering invalid objects first prevents the other filters from erroring due to an invalid object.

File: pkg/object/validator/policy.go
Comment (karlkfi) [reply pkg/object/validator/policy.go:7]: > So it seems like SkipInvalid is the approach that best fits our overall approach to errors here.

For now, yes. Tho in the future I'd like to add change groups so we can continue on error by default, but skip whole change groups if any of their objects are invalid. With that behavior, if you don't specify any change groups, all resources are considered part of the same change group, meaning that the default behavior would be equivalent to ExitEarly. Then if you want SkipInvalid behavior you just have to define multiple independent change groups. Or you could opt into SkipInvalid for all objects by default, equivalent to giving every object its own change group.

> I think there will be some issues that will prevent us from continuing

Yes. It will likely error in multiple places. I'll need to add tests for invalid input and make sure it handles them as desired.

> Also, circular dependencies isn't currently caught by the validation and it is not clear how we would skip resources in that situation.

Agreed. It's not obvious what the user's desired behavior would be in the case of circular dependencies either. I think for now we can just throw an error and exit and figure out how to skip it later, if needed.

> Could we simplify by making [SkipInvalid] the behavior until we have evidence that something else is needed?

@seans3 suggested using SkipInvalid as the default behavior. I could be convinced to do it that way, but I explained above why I think ExitEarly is a more forwards compatible default. 

IME building continuous deployment pipelines, I have always in the past preferred exiting early and stopping deployment if there's any problem that needs human input. Validation errors are almost always user mistakes, and partial deployment can easily cause production outage. So if given the option as a user, I would choose ExitEarly behavior. 

That said, I know some users have asked for continue-on-error, and especially after you've already started applying, it's not obvious if continuing or stopping would cause more brokenness. But I suspect that if we presented these same users with the change group option I suggested, they would prefer that over just continue-on-error behavior.

> I think ExitEarly do have some good use-cases, but I'm uncertain if this is the right way to do it.

This is the current behavior. Changing it would violate the principle of least surprise for users. Plus, Config Sync already does it's own validation earlier and would exit early anyway. It just doesn't validate all the same things, like the dependency annotations.

> For most of the validation, we want to let users shift it left and detect these issues in their CI/CD pipelines rather than when they are ready to apply. So it seems like we want this to be available outside of a pre-step to actually applying it. Also, we already provide a dry-run flag, which should detect these issues as well as many additional validation errors.

IME, dry-run is only really useful in a few cases:
1. When applying manually and you want to be see a diff of the changes to make sure hydration didn't have any unexpected side effects.
2. When there are multiple clients modifying the same files and you want to make sure you're not reverting someone else's changes accidentally.
3. When applying with automation and you want a record of the diff to be stored for debugging later or the diff needs manual approval.

There's not really anything to run between dry-run and apply that makes it useful to do a dry-run in CI/CD otherwise, is there? And in all cases, you'd still want to validate again in the apply, in case they didn't run dry-run, right? 

In any case, this validation code is the same code being used for both the apply and the dry-run, and I'm pretty sure we want the options and defaults to be the same for both, right?

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go:149]: AFAIK, skipped objects aren't added or removed from the inventory. If they were previously in the inventory, I think they're retained, right?

So is there a problem?

File: pkg/object/graph/depends.go
Comment (karlkfi) [reply pkg/object/graph/depends.go:100]: I think I can make SortObjs return a list of validation errors, without exiting on each one. Then we can move the SortObjs call from the solver to the applier/destroyer and send the ValidationEvents from there.

**Review by seans3 — COMMENTED** [review]
I believe this approach to validation is misguided for a couple reasons:

* We should emphasize `--dry-run` as the primary validation mechanism. Instead of attempting to figure out problems ahead of time, `--dry-run` runs as much of the entire apply algorithm as possible, surfacing errors throughout the algorithm. This type of validation will almost always be better than pre-run client-side validation. It looks like the validations added here for `depends-on` and `apply-time mutation` are already implemented as errors in `--dry-run`. I do not see much utility for these validations other than better error messages.
* Exiting early will be especially problematic, since this makes significant assumptions about what the library client wants and differs dramatically from the underlying `kubectl apply` which continues applying when encountering an apply error. Exiting early will probably require manually calling the final inventory set task to ensure the prune/inventory algorithm continues to work.
* If a library clients want stronger validation (or domain-specific validation), they can always implement it themselves before calling the library apply. This library has to work for all clients, and exiting early makes strong assumptions.

I believe the goal of client-side validation should be much more modest, such as generating a more informative error messages. If the history of Kubernetes CLI's is any guide, over-ambitious client-side validation will be jettisoned eventually for dry-run and server-side validation. But as long as there is a validation option other than `Exit Early`, I will not object to this PR.

**Comment by karlkfi** [comment]
@seans3 The current existing behavior already exits early if the Validator errors. I don't understand why you think that's misguided. The code is littered with hundreds of places that assume that minimal validation has been done. It's not going to be a small change to fix that to skip invalid objects (which seems to be what you want).

What this PR primarily does is adds the ability to skip those invalid objects. 

I also added more explicit depends-on & apply-time-mutation errors, but that's because these are client-side annotations and the server won't validate them. So if they're invalid and the applier continues without skipping them (the current behavior), then that's behaviorally incorrect. We should never completely ignore invalid depends-on & apply-time-mutation errors. We should either skip those invalid objects or exit early. Either way, the user needs to be notified. That's why I added the ValidationError, to make sure the user always gets notified, regardless of whether they chose SkipInvalid or ExitEarly.

Now that I've said that tho, I think the ApplyAll policy is probably only useful when not using depends-on or apply-time-mutation. Otherwise the behavior when they're invalid would be obviously incorrect. ApplyAll would basically be opting out of client-side validation and ignoring errors in client-side features. That's basically what you're championing when you're saying the applier should be more like kubectl and do less validation. The big difference is that kubectl doesn't have any client-side-only features like depends-on or apply-time-mutation.

File: pkg/object/validator/validate.go
Comment (karlkfi) [reply pkg/object/validator/validate.go]: Alright, I've moved all the dependency and sorting errors to SortObjs and moved that forward in the applier so we can use the errors as validation errors, and then modified SortObjs to return a MultiError so we get more than one at a time.

File: pkg/object/validator/validate.go
Comment (karlkfi) [reply pkg/object/validator/validate.go]: I've moved this to SortObjs and comments on the why,.

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go]: I've added invalid objects to the TaskContext now. Invalid objects will only be added to the inventory IFF they already existed in the inventory. So that can only really happen for annotations that were changed after being applied, and not for missing name/kind/namespace.

**Comment by karlkfi** [comment]
I've made significant changes to the handling of dependency validation. Please re-review.

- Move dependency validation from Validator to SortObjs.
- Move SortObjs call from solver to applier/destroyer.
  This allows the resulting errors to be treated as validation errors.
- Modify SortObjs to return a MultiError so that all validation errors
  can be sent and invalid objects can be skipped or cause early exit.
- Add invalid objects to the TestContext so they can be retained in
  the inventory (only if already present). This primarily applies to
  invalid annotations and dependencies. Objects without name or kind
  should never be added to the inventory.
- Remove invalid objects from the applySet/pruneSet, rather than
  filtering and skipping them in the apply/prune/wait tasks. This
  reduces the number of events that are redundant now with validation
  events.
- Handle CyclicDependencyError as a validation error. It applies to
  multiple objects, instead of just one, like ValidationError.
- Handle validation of dependency errors in destroyer as well as
  the applier.
- Replace many -OrDie object commands that were panicing when an
  invalid object was found. Add ValidateObjMetadata to do the
  validation that the -OrDie functions were doing.
- Replace MultiValidationError with a more generic MultiError.
- Simplify ValidationError to optionally wrap MultiError.
- Modify ValidationError output to be easier to read.
- Add e2e test for invalid object handling.

I still have more tests to add.

File: pkg/apply/event/event.go
Comment (karlkfi) [inline pkg/apply/event/event.go:208]: Most of the other event operations are prefixed by something consistent.

Do we want to change these to use Applied/Apply as prefix?

It would make them easier to auto-complete. We can still print them as something without the prefix, because the generator allows specifying the string value as an inline comment.

File: pkg/apply/applier.go
Comment (karlkfi) [inline pkg/apply/applier.go:191]: This will break waiting for dependencies if the dependency is invalid, but only if SkipInvalid is enabled. Since ExitEarly is default for reverse compatibility, this is not technically a regression. The intended fix will be a new feature to filter objects whose dependencies aren't reconciled. The dependency filter is discussed in https://github.com/kubernetes-sigs/cli-utils/issues/455

File: pkg/apply/applier.go
Comment (karlkfi) [inline pkg/apply/applier.go:193]: This new pre-filtering might make the InvalidObjectFilter obsolete. If the objects never get added to the apply/prune tasks, they don't need to be filtered then.

File: pkg/apply/event/event.go
Comment (karlkfi) [reply pkg/apply/event/event.go:208]: This should probably be punted to a different PR. It will technically break reverse compatibility by changing an existing event.

File: pkg/object/objmetadata_set.go
Comment (karlkfi) [inline pkg/object/objmetadata_set.go:142]: This hash change can probably be extracted to a different PR. It's needed for this PR, but can be made standalone.

File: pkg/object/unstructured.go
Comment (karlkfi) [inline pkg/object/unstructured.go:34]: Even with the changes in this PR, there are still 127 usages of UnstructuredToObjMetaOrDie in this repo. 

A lot of them are in tests and can probably be replaced by UnstructuredToObjMetadata, but the rest are in the code and could cause a panic if executed on a code path during a SkipInvalid apply/destroy.

Comment (karlkfi) [reply pkg/object/unstructured.go:34]: I'm not sure if it's a good idea to try to replace them all in this PR or not.

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go:193]: removed InvalidObjectFilter

File: pkg/apply/event/event.go
Comment (karlkfi) [reply pkg/apply/event/event.go]: Punted to https://github.com/kubernetes-sigs/cli-utils/issues/502

**Comment by karlkfi** [comment]
Extracted replacement of -OrDie (panic) functions to https://github.com/kubernetes-sigs/cli-utils/pull/505

This should vastly reduce the number of files changed in this PR, once rebased.

File: pkg/object/unstructured.go
Comment (karlkfi) [reply pkg/object/unstructured.go]: This refactor was moved to https://github.com/kubernetes-sigs/cli-utils/pull/505

**Comment by karlkfi** [comment]
This has been rebased on top of https://github.com/kubernetes-sigs/cli-utils/pull/505 and https://github.com/kubernetes-sigs/cli-utils/pull/506

File: pkg/object/validation/error.go
Comment (mortent) [inline pkg/object/validation/error.go:13]: It would be useful to explain the reason for this interface. It looks like `verror` is the only implementation, so why do we need the interface?

**Review by mortent — COMMENTED** [review]
I've done a first pass. The direction of allowing users to continue even if a subset of the resources have issues is aligned with the direction we are going.

I know some code has already been moved out of this PR, but I'm still finding it hard to do a good review since there is a lot going on. Is there a way to move more out into a separate PR? In particular, the changes to the `pkg/object/graph` package. Also maybe some of the additional validation error changes could be added in a separate PR prior to adding the ValidationEvent changes.

File: pkg/object/validation/error.go
Comment (karlkfi) [reply pkg/object/validation/error.go:31]: oops. the ValidationError struct is obsolete. It's replaced by the validation.Error interface, but got dragged along in a rebase. I'll delete it.

File: pkg/object/validation/error.go
Comment (karlkfi) [reply pkg/object/validation/error.go:13]: I can probably get rid of the interface. I thought it would be useful, but all it ends up doing is avoiding a `*` when casting.

Technically the Unwrap function is a new Go standard, but they didn't make a public interface for it when they added it to the logger. So it doesn't actually need a parent interface to compose it.

File: pkg/object/validation/validate.go
Comment (karlkfi) [reply pkg/object/validation/validate.go:18]: yes, by the Validation func.

Comment (karlkfi) [reply pkg/object/validation/validate.go:18]: nope. rebase error. replaced by MultiError

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go:184]: Technically, no. But it makes the new default ExitEarly mode reverse compatible without needing the client to watch for ValidationEvents.

File: pkg/apply/solver/solver_test.go
Comment (karlkfi) [reply pkg/apply/solver/solver_test.go:105]: Probably not, except as example of how to use the Solver when dealing with dependencies. These are what the inputs will look like now that sorting has been externalized.

File: pkg/apply/destroyer.go
Comment (karlkfi) [reply pkg/apply/destroyer.go:146]: They're called deletes in the destroyer and prunes in the applier. I wish they were all just called deletes...

I suppose I could try wrapping this duplicate chunk in a function.

**Comment by karlkfi** [comment]
Extracted the MultiError changes to https://github.com/kubernetes-sigs/cli-utils/pull/508

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go:185]: Occurs to me that this could make for redundant printed output, since the printers print the validation events and error events.

File: pkg/apply/destroyer.go
Comment (karlkfi) [reply pkg/apply/destroyer.go:146]: I found a way to abstract some of the code into a `validateAndSort` function. It sounds weird, but sorting can produce validation errors.

File: pkg/object/validator/policy.go
Comment (karlkfi) [reply pkg/object/validator/policy.go]: I've removed ApplyAll for now. It's not actually possible to apply resources without a kind or name. So the feature wouldn't make much sense.

I've left ExitEarly as the default, because that's the current behavior. We can change this later, if we want.

This PR primarily adds SkipInvalid and makes it opt-in.

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go]: I pulled out some of the logic to a `validateAndSort` function.

> The original design was that more of of the logic for determining which resources should be applied/pruned and the order was in the Solver type.

Correct. This PR changes that, which makes the solver less of a solver. Another alternative would be to make the solver return a list of validation errors. Would that be preferable?

File: pkg/apply/applier_test.go
Comment (karlkfi) [reply pkg/apply/applier_test.go]: Fixed.

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go]: I've changed the behavior to not send ValidationEvents when configured to ExitEarly.

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go]: Ok, I moved the sorting back into the Solver. It's slightly cleaner and fewer changes. To make it work, i had to inject InvalidIds into the Solver, because SortObjs needs the full input set, and then invalid objects need to be filtered afterwards, before the Task is created.

One interesting side effect of this is that the InvAddTask WILL include "invalid" objects that fail during sorting (depends-on, apply-time-mutation, and cycle errors), because the InvAddTask is built before sorting happens. This is actually exactly the same behavior as the current code. So it's not a regression. If we want to change this behavior, we can make the InvAddTask do filtering using the InvalidIds in the TaskContext.

File: pkg/apply/destroyer.go
Comment (karlkfi) [reply pkg/apply/destroyer.go]: Duplication should be reduced now that the sorting is moved back into the Solver.

File: pkg/apply/solver/solver_test.go
Comment (karlkfi) [reply pkg/apply/solver/solver_test.go]: These are no longer duplicates with sorting moved back into the Solver. So I reverted this file and added some error details.

**Comment by karlkfi** [comment]
Ok, I did a big refactor to move the sorting bakc into the Solver. And got the tests finally passing again.

TODO: 
- extract the taskrunner refactor to inject the TaskContext
- extract the Hash change to not return an error
- extract the Sorting fixes
- add test case for InvSetTasks filtering invalid objects

**Comment by karlkfi** [comment]
/unhold

All TODOs finished. Ready for another review pass.

File: pkg/apply/task/inv_set_task.go
Comment (karlkfi) [reply pkg/apply/task/inv_set_task.go:110]: Added

File: pkg/apply/solver/solver.go
Comment (karlkfi) [inline pkg/apply/solver/solver.go:108]: Should we replace this pre-filtering with the original InvalidObjectFilter instead?

**Review by mortent — APPROVED** [review]
I think this looks good. I see that we have had some discussions on this PR about when validation should happen, but the primary feature here is that we can continue applying resources in many more situations, which I think is a direction we all agree on. Allowing early exit seems reasonable to support, not least because that is the current behavior.

/approve

**Comment by mortent** [comment]
I'll remove the approval just to make sure @seans3 also has a chance to look at it before it merges.
/approve cancel

File: pkg/apply/solver/solver.go
Comment (karlkfi) [reply pkg/apply/solver/solver.go:108]: Extracted filtering to validation.Collector

File: pkg/apply/applier.go
Comment (karlkfi) [reply pkg/apply/applier.go]: Why is that better than type comparison?

Comment (karlkfi) [reply pkg/apply/applier.go]: Oh, i see. errors. As will recursively Unwrap errors until it finds the specified type. That makes sense, in case any of the validation errors are wrapped. Unfortunately, it wont work if there's a MultiError in the stack. But in this case, I already unwrapped those before adding them to the Collector (this code is in the Collector now).

Comment (karlkfi) [reply pkg/apply/applier.go]: done

**Review by seans3 — CHANGES_REQUESTED** [review]
I do not believe exiting early should be the default behavior. Even if this behavior exists now, it is incorrect and should be fixed to continue on error. Exiting early is also not consistent with the `kubectl apply` behavior which continues on error.

**Comment by karlkfi** [comment]
> I do not believe exiting early should be the default behavior. Even if this behavior exists now, it is incorrect and should be fixed to continue on error. Exiting early is also not consistent with the kubectl apply behavior which continues on error.

It's likely that ConfigSync and kpt will both initially use ExitEarly when adopting this new code. They may later switch to either hard coding SkipInvalid OR piping the option through to the user. Does that affect your default suggestion @seans3 ?

**Comment by karlkfi** [comment]
FWIW, I do not consider ExitEarly to a be a bug. Its a feature which arises from the desire to group related changes. Of course, Kubernetes doesn't have atomic applies or groups, but you can at least avoid applying a Deployment if its Secret, ServiceAccount, or ConfigMap is invalid. This behavior is highly desired when you take the time to granularly group your objects.

On the other hand, SkipInvalid is attractive when you haven't or don't intend to group your objects. This comes up regularly when multiple teams share a single config repo or delegate config management to a central SRE/DevOps/Ops team.

I'm ok with either being default. I don't think it really matters. There are pros and cons on both sides.
- ExitEarly is reverse compatible, which is desirable for existing kpt users because they're releasing beta versions and breaking changes aren't being released as major versions yet.
- SkipInvalid might be more desirable to new kpt users who are more familiar with kubectl.
- Any wrapping client can select their own default... So as long as we communicate breaking changes, I'm not against them.

I'm biasing slightly towards reverse compatibility here, but if anyone else wants to chime in with a 3rd vote I don't mind making the change.

**Comment by mortent** [comment]
I would probably prefer that we keep the current behavior here (so ExitEarly is the default), but I don't feel strongly about it.

**Comment by haiyanmeng** [comment]
For Config Sync users, the default should be ExitEarly, which is closer to the existing validation behavior in Config Sync.

**Review by seans3 — APPROVED** [review]
/lgtm

**Pull request merged**
